### PR TITLE
[CORE-10976] Add number of connections limiting QoS control

### PR DESCRIPTION
**Pull request by coutinhop**
## Description

Add limiting of the number of ingress/egress connections via the workload's QoSControls (ultimately translated into iptables/nftables rules).

## Todos

- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
Add packet number of connections limiting QoS control to workloads.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/fv/qos_controls_test.go [hunk 1]
```diff
@@ -289,6 +298,185 @@ var _ = infrastructure.DatastoreDescribe(
...
+					Expect(err).NotTo(HaveOccurred())
+				}
+
+				By("Killing and cleaning up nginx server process")
+				err = serverCmd.Process.Kill()
+				Expect(err).NotTo(HaveOccurred())
+				err = serverCmd.Process.Release()
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
```
Comment (nelljerram) [inline felix/fv/qos_controls_test.go:479]: This is great testing, but I think it would be worth seeing what common functions we can define across the whole of the QoS testing, so that the tests become more easily comprehensible overall.  I've suggested one above, which I think is a really easy one.  Beyond that, I don't think further work is strictly needed for this PR, but I do think that it would help longer term.  The next place I'd look would be the repeating pattern for setting a QoSControl and then waiting for that to be reflected in iptables or TC.

**Review by nelljerram — DISMISSED** [review]
Great to develop this so quickly!  But I think the new iptables rule isn't in the best place.

File: felix/docker-image/Dockerfile [hunk 2]
```diff
@@ -52,7 +52,9 @@ RUN apt-get update && apt-get install -y \
     hping3 \
     libpcap0.8 \
     tcpdump \
-    iperf3
+    iperf3 \
+    nginx \
+    netcat-traditional
```
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/docker-image/Dockerfile:57]: yes, I believe the comment is outdated, I had spoken to @fasaxc and this image is currently only used in testing

File: felix/rules/endpoints_test.go [hunk 3]
```diff
@@ -1465,6 +1465,73 @@ var _ = Describe("Endpoints", func() {
...
+					{
+						Name: "cali-tw-cali1234",
+						Rules: []generictables.Rule{
+							// ingress max connection rules
+							{
+								Match:  Match(),
+								Action: LimitNumConnectionsAction{Num: 20, With: generictables.RejectWithTCPReset},
+
+								Comment: []string{"Reject connections over ingress connection limit"},
+							},
```
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/rules/endpoints_test.go:1489]: I will double check this, but I had assumed that this rule needed to look at all connections to keep track of its state, but if it doesn't, then yes it's best to move the rule around

File: felix/iptables/actions.go [hunk 4]
```diff
@@ -436,3 +443,20 @@ func (a LimitPacketRateAction) ToFragment(features *environment.Features) string
...
+	Num                     int64
+	With                    generictables.RejectWith
+	TypeLimitNumConnections struct{}
+}
+
+func (a LimitNumConnectionsAction) ToFragment(features *environment.Features) string {
+	if a.Num < 0 {
+		logrus.WithField("rate", a.Num).Panic("Invalid limit")
+	}
+	return fmt.Sprintf("-p tcp -m conntrack --ctstate NEW,RELATED,ESTABLISHED -m connlimit --connlimit-above %d --connlimit-mask 0 -j REJECT --reject-with %s", a.Num, a.With)
```
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/iptables/actions.go:457]: from iptables docs, the only ctstates not mentioned are SNAT and DNAT. I'm not sure those apply, but will look into it

File: felix/fv/qos_controls_test.go [hunk 5]
```diff
@@ -289,6 +298,185 @@ var _ = infrastructure.DatastoreDescribe(
...
+			It("should limit connections correctly", func() {
+				const numConnections = 4
+				var netcatCmds [numConnections]*exec.Cmd
+
+				By("Starting nginx server on workload 0")
+				serverCmd := w[0].ExecCommand("nginx")
+				err := serverCmd.Start()
+				Expect(err).NotTo(HaveOccurred())
+				// Sleep to give nginx some time to start up
+				time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/fv/qos_controls_test.go:319]: good point, will do!

File: felix/rules/endpoints_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/rules/endpoints_test.go:1489]: Update: you're right, I was able to move the rules with no detriment to the expected functionality in the tests, thanks for catching it!

File: felix/fv/qos_controls_test.go [hunk 6]
```diff
@@ -289,6 +298,185 @@ var _ = infrastructure.DatastoreDescribe(
...
+				} else {
+					// ingress config should be present
+					Eventually(getRules(0), "10s", "1s").Should(MatchRegexp(`-A cali-tw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over ingress connection limit" -m conntrack --ctstate NEW,RELATED,ESTABLISHED -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+					// egress config should not be present
+					Consistently(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-fw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over egress connection limit" -m conntrack --ctstate NEW,RELATED,ESTABLISHED -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+				}
+
+				By("Starting persistent netcat connections on workload 1")
```
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/fv/qos_controls_test.go:360]: I ended up only partially doing that, with a function to start the persisent connections, because at one of the test cases I added testing for the edge case we talked about offline (closing one of the initial persistent connections to free up the limit and checking that eventually the `wget` command does succeed)

File: felix/docker-image/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (coutinhop) [reply felix/docker-image/Dockerfile:57]: will do!

[the remark this answers is not in the document]
Comment (fasaxc) [reply felix/docker-image/Dockerfile:54]: Yes, it's only a test image.  +1 for renaming it.

The comment about keeping it in sync with node is still relevant though; we still want to make sure that it has a similar base set-up to minimize test invalidation.

File: felix/fv/qos_controls_test.go [hunk 7]
```diff
@@ -289,6 +299,180 @@ var _ = infrastructure.DatastoreDescribe(
...
+					// egress config should not be present
+					Consistently(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`(?s)chain filter-cali-fw-` + w[0].InterfaceName + ` {.*\n.*meta l4proto tcp counter packets \d+ bytes \d+ ct count over 4 reject with tcp reset comment ".*Reject connections over egress connection limit"`))
+				} else {
+					// ingress config should be present
+					Eventually(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-tw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over ingress connection limit" -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+					// egress config should not be present
+					Consistently(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-fw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over egress connection limit" -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+				}
```
Comment (fasaxc) [inline felix/fv/qos_controls_test.go:399]: Generally best to avoid asserting on individual rules in FVs; tend to make the tests brittle vs code changes. I suppose here you're trying to make sure the dataplane has settled before you proceed?  I guess that's OK, but I might trim the regex down to something minimal `--connlimit-saddr` or something like that.  Just so it gives a clue that the rule has disappeared without duplicating UT-level rule checking.

**Review by fasaxc — CHANGES_REQUESTED** [review]
As discussed, we already have `workload.StartPersistentConnection`; would be good to double down on that if we can.

File: felix/fv/qos_controls_test.go (hunk 7 above)
Comment (coutinhop) [reply felix/fv/qos_controls_test.go:391]: hm, yes the intention was to make sure the dataplane has settled before moving on... I've done this on the packet rate test case as well, will adjust this file where it applies

**Review by fasaxc — APPROVED** [review]
